### PR TITLE
Update Invitation

### PR DESCRIPTION
**Pull request by alejandro-castellon**
[AB#337](https://dev.azure.com/funjaladev/7867a1c5-5ac7-4c7f-baff-87d740bf4fc1/_workitems/edit/337)

# Overview

- Implement API controller
- Implement invitation validator
- Implement update service
- Implemet mongo and memory dal
- Add unit tests

### CODE REVIEW

**Review by paulo-teran-jalafund — DISMISSED** [review]

**Review by lina-ramirez-jalafund — DISMISSED** [review]

File: src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    using System;
+    using FluentValidation;
+    using Jalasoft.TeamUp.Projects.Models;
+
+    public class InvitationValidator : AbstractValidator<Invitation>
+    {
+        public InvitationValidator()
+        {
+            this.RuleFor(invitation => invitation.Id)
+                .NotEmpty().NotNull();
```
Comment (cristian-chavez-jalafund) [inline src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs]:
```suggestion
                .NotEmpty();
```
is enough with .NotEmpty() only
https://docs.fluentvalidation.net/en/latest/built-in-validators.html#notempty-validator

File: src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+            this.RuleFor(invitation => invitation.ProjectName)
+                .Length(3, 15)
+                .Matches("^[a-zñ A-ZÑ]+$")
+                .NotEmpty().NotNull();
+
+            this.RuleFor(invitation => invitation.ResumeName)
+                .Length(3, 15)
+                .Matches("^[a-zñ A-ZÑ]+$")
+                .NotEmpty().NotNull();
```
Comment (cristian-chavez-jalafund) [inline src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs]: this should be first IMO

File: src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+            this.RuleFor(invitation => invitation.TextInvitation)
+                .MaximumLength(160);
+
+            this.RuleFor(invitation => invitation.Status)
+                .Matches("^[a-zñ A-ZÑ]+$");
```
Comment (cristian-chavez-jalafund) [inline src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs]: maybe an enumerable should be considered to have a defined list of possible statuses?

File: src/dal/Jalasoft.TeamUp.Projects.DAL/InvitationsInMemoryRepository.cs [hunk 4]
```diff
@@ -60,17 +60,22 @@ public IEnumerable<Invitation> GetAllInvitationsByResumeId(int resumeId)
...
         public void Remove(Guid id)
         {
             throw new NotImplementedException();
         }
 
-        public Invitation UpdateById(Invitation project)
+        public Invitation UpdateById(Invitation invitation)
         {
-            throw new NotImplementedException();
+            var invitationsList = new List<Invitation>(invitations);
```
Comment (cristian-chavez-jalafund) [inline src/dal/Jalasoft.TeamUp.Projects.DAL/InvitationsInMemoryRepository.cs:74]: I think there's no need to use this extra variable, all could be performed in `invitations` only

Comment (cristian-chavez-jalafund) [reply src/dal/Jalasoft.TeamUp.Projects.DAL/InvitationsInMemoryRepository.cs:74]: or maybe am I missing something? please double check

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            this.mockService = new Mock<IInvitationsService>();
+            this.mockHttpContext = new DefaultHttpContext();
+            this.updateInvitation = new PatchInvitationById(this.mockService.Object);
+        }
+
+        [Theory]
+        [InlineData(@"[
+        {""op"" : ""replace"", ""path"" : ""/status"", ""value"" : ""Accepted""}
+        ]")]
```
Comment (cristian-chavez-jalafund) [inline tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs:31]: if you have only one use case (inline data) this could be a Fact instead of a Theory
add more test cases (edge scenarios) which I'd do, or make this as Fact (which I'd avoid)

File: tests/core/Jalasoft.TeamUp.Projects.Core.Tests/InvitationsServiceTests.cs [hunk 6]
```diff
@@ -108,5 +109,20 @@ public void GetInvitationsByProjectId_InvitationsDontExist_EmptyArray()
...
+            this.mockRepository.Setup(repository => repository.GetById(Guid.Parse("5a7939fd-59de-44bd-a092-f5d8434584de"))).Returns(stubInvitation);
+
+            // Act
+            var result = this.service.UpdateInvitation(stubInvitation);
+
+            // Assert
+            Assert.IsType<Invitation>(result);
+        }
     }
 }
```
Comment (cristian-chavez-jalafund) [inline tests/core/Jalasoft.TeamUp.Projects.Core.Tests/InvitationsServiceTests.cs]: please add unit tests for `InvitationValidator`

File: src/api/Jalasoft.TeamUp.Projects.API/Controllers/PatchInvitationById.cs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        [OpenApiParameter(name: "id", In = ParameterLocation.Path, Required = true, Type = typeof(Guid), Description = "The invitation identifier.")]
+        [OpenApiRequestBody("application/json", typeof(JsonPatchDocument<Invitation>), Description = "JSON request body")]
+        [OpenApiResponseWithBody(statusCode: HttpStatusCode.OK, contentType: "application/json", bodyType: typeof(Invitation), Description = "Successful response")]
+        [OpenApiResponseWithoutBody(statusCode: HttpStatusCode.NotFound, Description = "Resource not found")]
+        public async Task<IActionResult> UpdateInvitation(
+            [HttpTrigger(AuthorizationLevel.Anonymous, "patch", Route = "v1/invitations/{id}")] HttpRequest req, Guid id)
+        {
+            try
+            {
+                var invitation = this.updateInvitationService.GetInvitation(id);
```
Comment (Raul-Gamarra-jalafund) [inline src/api/Jalasoft.TeamUp.Projects.API/Controllers/PatchInvitationById.cs:41]: since you return an invitation from the invitations service you could validate that object to determine the existence of the invitation, that way you don't need to get the invitation

File: src/core/Jalasoft.TeamUp.Projects.Core/InvitationsService.cs [hunk 8]
```diff
@@ -25,5 +27,18 @@ public Invitation[] GetInvitationsByResumeId(int resumeId)
...
+        {
+            return this.invitationsRepository.GetById(invitationId);
+        }
+
+        public Invitation UpdateInvitation(Invitation invitation)
+        {
+            InvitationValidator validator = new InvitationValidator();
+            validator.ValidateAndThrow(invitation);
+            var result = this.invitationsRepository.UpdateById(invitation);
+            return invitation;
```
Comment (Raul-Gamarra-jalafund) [inline src/core/Jalasoft.TeamUp.Projects.Core/InvitationsService.cs]: you could be returning the actual result instead of the input parameter

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        public async void PatchInvitationById_IdIsValid_OkObjectResult(string body)
+        {
+            Invitation stubInvitation = StubInvitation.GetStubInvitation();
+
+            HttpRequest request = this.mockHttpContext.Request;
+            request.Body = SetStream.Setstream(body);
+            this.mockService.Setup(service => service.GetInvitation(new Guid("5a7939fd-59de-33bd-a092-f5d8434584de"))).Returns(stubInvitation);
+            this.mockService.Setup(service => service.UpdateInvitation(null)).Returns(stubInvitation);
```
Comment (Raul-Gamarra-jalafund) [inline tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs]: why setting up this use case here? seems like you are not using it

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+        [Theory]
+        [InlineData(@"[
+        {""op"" : ""replace"", ""path"" : ""/status"", ""value"" : ""Accepted""}
+        ]")]
+        public async void PatchInvitationById_IdIsNotValid_NotFoundResult(string body)
+        {
+            HttpRequest request = this.mockHttpContext.Request;
+            request.Body = SetStream.Setstream(body);
+            this.mockService.Setup(service => service.GetInvitation(Guid.Parse("5a7939fd-59de-44bd-a092-f5d8434584de"))).Throws(new ProjectsException(ProjectsErrors.NotFound));
```
Comment (Raul-Gamarra-jalafund) [inline tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs]: the service does not return an exception when the id is not found it only returns null, the exception is handled in the API

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+            this.mockService.Setup(service => service.UpdateInvitation(null)).Throws(new ProjectsException(ProjectsErrors.NotFound));
```
Comment (Raul-Gamarra-jalafund) [inline tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs]: why you need this one here?

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+            var response = await this.updateInvitation.UpdateInvitation(request, new Guid("5a7939fd-59de-33bd-a092-f5d8434584de"));
+            var result = Assert.IsType<ProjectsErrors>(ProjectsErrors.NotFound);
+
+            Assert.IsType<ProjectsErrors>(ProjectsErrors.NotFound);
```
Comment (Raul-Gamarra-jalafund) [inline tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs]: please validate the status code also in this test

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        [InlineData(@"[
+        {""op"" : ""replace"", ""path"" : "" /statuuus"", ""value"" : ""Accepted""}
+        ]")]
+        public async void PatchInvitationById_BadRequestBody_BadRequestResult(string body)
+        {
+            Invitation stubInvitation = StubInvitation.GetStubInvitation();
+
+            HttpRequest request = this.mockHttpContext.Request;
+            request.Body = SetStream.Setstream(body);
+            this.mockService.Setup(service => service.GetInvitation(new Guid("5a7939fd-59de-33bd-a092-f5d8434584de"))).Returns(stubInvitation);
```
Comment (Raul-Gamarra-jalafund) [inline tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs:87]: is this case we could return a validation exception that use case CAN happen in the service

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+            this.mockService.Setup(service => service.UpdateInvitation(null)).Returns(stubInvitation);
+            var response = await this.updateInvitation.UpdateInvitation(request, new Guid("5a7939fd-59de-33bd-a092-f5d8434584de"));
+            var result = Assert.IsType<ProjectsErrors>(ProjectsErrors.BadRequest);
```
Comment (Raul-Gamarra-jalafund) [inline tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs]: you need to assert the response here

File: src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs (hunk 1 above)
Comment (alejandro-castellon) [reply src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs]: done

File: src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs (hunk 2 above)
Comment (alejandro-castellon) [reply src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs]: done

File: src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs (hunk 3 above)
Comment (alejandro-castellon) [reply src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs]: .Matches("^(Invited|Rejected|Accepted)$");
This could be a better way

File: tests/core/Jalasoft.TeamUp.Projects.Core.Tests/InvitationsServiceTests.cs [hunk 15]
```diff
@@ -108,5 +109,20 @@ public void GetInvitationsByProjectId_InvitationsDontExist_EmptyArray()
...
             Assert.Empty(result);
         }
+
+        [Fact]
+        public void PatchInvitationByProjectId_InvitationIsValid_InvitationUpdated()
+        {
+            // Arrange
+            Invitation stubInvitation = StubInvitation.GetStubInvitation();
+
+            this.mockRepository.Setup(repository => repository.GetById(Guid.Parse("5a7939fd-59de-44bd-a092-f5d8434584de"))).Returns(stubInvitation);
```
Comment (Raul-Gamarra-jalafund) [inline tests/core/Jalasoft.TeamUp.Projects.Core.Tests/InvitationsServiceTests.cs]: this should be mocking the update invitation method

File: tests/core/Jalasoft.TeamUp.Projects.Core.Tests/InvitationsServiceTests.cs [hunk 16]
```diff
@@ -47,7 +48,7 @@ public static IEnumerable<Invitation> MockInvitations()
                     TextInvitation = "We invite you to collaborate with the development team",
                     StartDate = DateTime.Today.AddDays(-10),
                     ExpireDate = DateTime.Today.AddDays(+10),
-                    Status = "invited"
+                    Status = "Invited"
```
Comment (Raul-Gamarra-jalafund) [inline tests/core/Jalasoft.TeamUp.Projects.Core.Tests/InvitationsServiceTests.cs:51]: please create a dictionary in the models that includes all the statuses so we can use it in this kind of situations, this way the developer will know that type of the statuses we have and wouldn't need to determine the write the exact name (this can easily cause mistakes)

File: tests/core/Jalasoft.TeamUp.Projects.Core.Tests/InvitationsServiceTests.cs (hunk 6 above)
Comment (alejandro-castellon) [reply tests/core/Jalasoft.TeamUp.Projects.Core.Tests/InvitationsServiceTests.cs]: done

File: src/dal/Jalasoft.TeamUp.Projects.DAL/InvitationsInMemoryRepository.cs (hunk 4 above)
Comment (alejandro-castellon) [reply src/dal/Jalasoft.TeamUp.Projects.DAL/InvitationsInMemoryRepository.cs:74]: We use this extra variable to have the methods Remove and Add

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs (hunk 5 above)
Comment (alejandro-castellon) [reply tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs:31]: done

File: tests/core/Jalasoft.TeamUp.Projects.Core.Tests/InvitationsServiceTests.cs (hunk 16 above)
Comment (cristian-chavez-jalafund) [reply tests/core/Jalasoft.TeamUp.Projects.Core.Tests/InvitationsServiceTests.cs:51]: I'd suggest an enumerable

File: src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+            this.RuleFor(invitation => invitation.ProjectName)
+                .Length(3, 15)
+                .Matches("^[a-zñ A-ZÑ]+$")
+                .NotEmpty();
+
+            this.RuleFor(invitation => invitation.TextInvitation)
+                .MaximumLength(160);
+
+            this.RuleFor(invitation => invitation.Status)
+                .Matches("^(Invited|Rejected|Accepted)$");
```
Comment (cristian-chavez-jalafund) [inline src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs]: use an enumerator instead of string values

File: src/api/Jalasoft.TeamUp.Projects.API/Controllers/PatchInvitationById.cs (hunk 7 above)
Comment (alejandro-castellon) [reply src/api/Jalasoft.TeamUp.Projects.API/Controllers/PatchInvitationById.cs:41]: We need to get the invitation to have the actual information plus the information that we want to send

File: src/core/Jalasoft.TeamUp.Projects.Core/InvitationsService.cs (hunk 8 above)
Comment (alejandro-castellon) [reply src/core/Jalasoft.TeamUp.Projects.Core/InvitationsService.cs]: done

File: src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs (hunk 17 above)
Comment (alejandro-castellon) [reply src/core/Jalasoft.TeamUp.Projects.Core/Validators/InvitationValidator.cs]: done

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs (hunk 9 above)
Comment (alejandro-castellon) [reply tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs]: done

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs (hunk 10 above)
Comment (alejandro-castellon) [reply tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs]: done

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs (hunk 11 above)
Comment (alejandro-castellon) [reply tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs]: done

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs (hunk 12 above)
Comment (alejandro-castellon) [reply tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs]: done

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs (hunk 13 above)
Comment (alejandro-castellon) [reply tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs:87]: done

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs (hunk 14 above)
Comment (alejandro-castellon) [reply tests/api/Jalasoft.TeamUp.Projects.API.Tests/PatchInvitationByIdTests.cs]: done

File: tests/core/Jalasoft.TeamUp.Projects.Core.Tests/InvitationsServiceTests.cs (hunk 15 above)
Comment (alejandro-castellon) [reply tests/core/Jalasoft.TeamUp.Projects.Core.Tests/InvitationsServiceTests.cs]: done

File: tests/core/Jalasoft.TeamUp.Projects.Core.Tests/InvitationsServiceTests.cs (hunk 16 above)
Comment (alejandro-castellon) [reply tests/core/Jalasoft.TeamUp.Projects.Core.Tests/InvitationsServiceTests.cs:51]: It is corrected with enumerable

**Review by cristian-chavez-jalafund — APPROVED** [review]

**Review by paulo-teran-jalafund — APPROVED** [review]

**Review by lina-ramirez-jalafund — APPROVED** [review]
